openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3248,164 @@ def from_openeye(oemol, allow_undefined_stereo=False):
+                                                            bond_stereochemistry_matching=False,)
+
+        if mapping is not None:
+            new_mol = offmol.remap(mapping)
+
+            # the pdb conformer is in the correct order so just attach it here
+            new_mol.add_conformer(pdbmol.conformers[0])
+
+            return new_mol
+
+        else:
+            raise InvalidConformerError('The PDB and SMILES structures do not match.')
+
+    def remap(self, mapping_dict, new_to_old=True):
+        """
+        Remap all of the indexes in the molecule to match the given mapping dict
+        :param mapping_dict: A dictionary of the mapping between in the indexes
+        :param new_to_old: The dict is {new_index: old_index} if True else {old_index: new_index}
+        :return:
+        """
+
+        # make two mapping dicts we need new to old for atoms
+        # and old to new for bonds
+        if new_to_old:
+            n_to_o = mapping_dict
+            o_to_n = dict((v, u) for u, v in mapping_dict.items())
+        else:
+            o_to_n = mapping_dict
+            n_to_o = dict((v, u) for u, v in mapping_dict.iteams())
+
+        new_molecule = Molecule()
+        new_molecule.name = self.name
+
+        # add the atoms list
+        for i in range(self.n_atoms):
+            # get the old atom info
+            old_atom = self._atoms[n_to_o[i]]
+            new_molecule.add_atom(atomic_number=old_atom.atomic_number,
+                                  formal_charge=old_atom.formal_charge,
+                                  is_aromatic=old_atom.is_aromatic,
+                                  stereochemistry=old_atom.stereochemistry,
+                                  name=old_atom.name)
+        # add the bonds
+        for bond in self._bonds:
+            a1 = o_to_n[bond.atom1_index]
+            a2 = o_to_n[bond.atom2_index]
+            new_molecule.add_bond(atom1=a1,
+                                  atom2=a2,
+                                  bond_order=bond.bond_order,
+                                  is_aromatic=bond.is_aromatic,
+                                  stereochemistry=bond.stereochemistry,
+                                  fractional_bond_order=bond.fractional_bond_order)
+
+        # remap the charges
+        new_charges = np.zeros(self.n_atoms)
+        for i in range(self.n_atoms):
+            new_charges[i] = self.partial_charges[n_to_o[i]].value_in_unit(unit.elementary_charge)
+        new_molecule.partial_charges = new_charges * unit.elementary_charge
+
+        # remap the conformers there can be more than one
+        if self.conformers is not None:
+            for conformer in self.conformers:
+                new_conformer = np.zeros((self.n_atoms, 3))
+                for i in range(self.n_atoms):
+                    new_conformer[i] = conformer[n_to_o[i]].value_in_unit(unit.angstrom)
+                new_molecule.add_conformer(new_conformer * unit.angstrom)
+
```

Comment:
Let's check whether there are any virtualsites, and raise a `NotImplementedError` if so

Reply:
Added this for now, but what would need to be done to allow us to remap the sites?

Reply:
We could remap atoms in the existing implementation of VirtualSites, but the existing implementation isn't functional in molecule parametrization, and will likely undergo serious refactoring. Currently the VirtualSite implementation is largely a placeholder.

If you'd like to implement support for the placeholder implementation of VirtualSites, which probably isn't worth the time, the biggest thing will be to make a copy of each original VirtualSite in the new molecule. Importantly, the new VirtualSite will need to have remapped `vsite.atoms` in the new molecule. Note that `vsite.atoms` isn't a list of atom _indices_, rather it is a _list of actual atoms_ (I did say it will undergo refactoring).

Reply:
okay, for now, I think it is best to just leave out the sites until the implementation is final.